File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem54_SuccessfulPairsOfSpellsAndPotions.java
package org.phoenix.leetcode.challenges;

import java.util.Arrays;

public class Problem54_SuccessfulPairsOfSpellsAndPotions {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n = spells.length;
        int m = potions.length;
        int[] pairs = new int[n];
        Arrays.sort(potions);
        for (int i = 0; i < pairs.length; i++) {
            int left = 0;
            int right = m - 1;
            int spell = spells[i];
            while (left <= right) {
                int mid = left + (right - left) / 2;
                long product = (long) spell * potions[mid];
                if (product >= success) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            pairs[i] = m - left;
        }

        return pairs;
    }
}

File: JavaScript/leetcode75/src/main/Problem54_SuccessfulPairsOfSpellsAndPotions.js
/**
 * @param {number[]} spells
 * @param {number[]} potions
 * @param {number} success
 * @return {number[]}
 */
var successfulPairs = function (spells, potions, success) {
  let n = spells.length;
  let m = potions.length;
  let pairs = [];
  potions.sort((a, b) => a - b);
  for (let i = 0; i < n; i++) {
    let left = 0;
    let right = m - 1;
    let spell = spells[i];
    while (left <= right) {
      let mid = Math.floor(left + (right - left) / 2);
      let product = spell * potions[mid];
      if (product >= success) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }
    pairs[i] = m - left;
  }
  return pairs;
};

module.exports = successfulPairs;
